exercises/edward_brown/ex01_basics/include/MakeWordFrequencyMap.h

```diff
@@ -0,0 +1,46 @@
+#ifndef EX1_MAKE_WORD_FREQUENCY_MAP
+#define EX1_MAKE_WORD_FREQUENCY_MAP
+#include "ForEachWord.h"
+#include "Punctuation.h"
+#include "ToLowerCase.h"
+#include <unordered_map>
+
+/** A predicate which determines if CandidateWord is a word for the purposes of
+ *  word frequency analysis.
+ *
+ *  @param CandidateWord The candidate to check.
+ *  @return True if CandidateWord is a 'word' false if not.
+ */
+bool isWord(std::string const &CandidateWord) {
+  auto constexpr MINIMUM_WORD_LENGTH = 4;
+  return CandidateWord.size() > MINIMUM_WORD_LENGTH;
+}
+
+/** Creates a map from strings containing words found in the range [Begin, End)
+ * to integer values denoting the corresponding number of occurences in said
+ * range.
+ *
+ *  This analysis is case and punctuation insensitive. Words with a length
+ * shorter than four characters are ignored.
+ *
+ *  @param Begin The const_iterator marking the beginning of the range to search
+ * for words.
+ *  @param End The const_iterator marking the end of the range to search for
+ * words.
+ *  @return The map built from the word frequency analysis performed.
+ */
+template <typename ConstCharInputIterator>
+std::unordered_map<std::string, int>
+makeWordFrequencyMap(ConstCharInputIterator begin, ConstCharInputIterator end) {
+  auto wordFrequencyMap = std::unordered_map<std::string, int>();
```

Comment:
Please see the earlier discussion on this PR (Punctuation.cpp) with @gemmaguest regarding the use of auto for *local* variables. It mostly comes down to ~~three~~ four things.

1. Less issues with the [most vexing parse](https://en.wikipedia.org/wiki/Most_vexing_parse). Which is especially useful when the LLVM style guide explicitly outlaws uniform initialization in the majority of cases. 
2. It's harder to end up with any uninitialized variables since `auto` variables must have an initializing expression.
3. There is less work and changes required when refactoring functions used to initialize variables or renaming the deduced (tracked) types themselves since they are deduced rather than written explicitly. 

4. When declaring multiple locals on consecutive lines, the variable names line up consistently which I feel improves readability but I accept may just be personal taste...

```
auto firstName = std::string("Joe");
auto surname = std::string("Bloggs");
auto age = 27;

// rather than

std::string firstName("Joe");
std::string surname("Bloggs");
int age = 21;
```
 However, obviously when dealing with non STL library types which may be expensive to move/copy I understand that some extra care must be taken and as was suggested above I am of course more than happy to "fit in with the local style where appropriate".